### PR TITLE
Reenable Transparent Compiler tests

### PR DESCRIPTION
**Pull request by majocha**
Experiment branched out from #17872

Async flavor of graph processing is simplified a bit and runs with Async.Parallel, given that it is async already.

If there are any remaining instabilities in Transparent Compiler tests they should be easier to weed out in a parallel test run, where the thread pool is way more stressed.

All the tests tests from #16766 but two should be unskipped, unless I missed some.

### CODE REVIEW

File: tests/FSharp.Compiler.Service.Tests/MultiProjectAnalysisTests.fs [hunk 1]
```diff
@@ -443,30 +443,36 @@ let z = Project1.x
             OtherOptions = Array.append options.OtherOptions [| ("-r:" + MultiProjectDirty1.dllName) |]
             ReferencedProjects = [| FSharpReferencedProject.FSharpReference(MultiProjectDirty1.dllName, MultiProjectDirty1.getOptions()) |] }
 
+// Because of writing to shared physical files
+// TODO: make the test cases independent
+[<FSharp.Test.RunInSequence>]
 [<Theory>]
+// Investigate: check count varies betweeen 5 and 6
 // [<InlineData(true)>]
```
Comment (majocha) [inline tests/FSharp.Compiler.Service.Tests/MultiProjectAnalysisTests.fs:451]: This is a single case that is genuinely unstable now, reproducible locally.

File: tests/FSharp.Compiler.Service.Tests/MultiProjectAnalysisTests.fs [hunk 2]
```diff
@@ -557,24 +562,23 @@ let ``Test multi project symbols should pick up changes in dependent projects``
...
     printfn "Old write time: '%A', ticks = %d"  wt1b wt1b.Ticks
     printfn "New write time: '%A', ticks = %d"  wt2b wt2b.Ticks
 
-    count.Value |> shouldEqual 4
+    count |> shouldEqual 4
     let wholeProjectResults2AfterChange2 = checker.ParseAndCheckProject(proj2options) |> Async.RunImmediate
 
-    System.Threading.Thread.Sleep(1000)
-    count.Value |> shouldEqual 6 // note, causes two files to be type checked, one from each project
+    count |> shouldEqual 6 // note, causes two files to be type checked, one from each project
```
Comment (majocha) [inline tests/FSharp.Compiler.Service.Tests/MultiProjectAnalysisTests.fs:576]: With Transparent Compiler the check count sometimes stops at 5.

Comment (majocha) [reply tests/FSharp.Compiler.Service.Tests/MultiProjectAnalysisTests.fs:595]: This should be now sorted out now. Apparently it had something to do with reusing the same physical files among test cases.

**Comment by majocha** [comment]
image
LruCache Count needs lock. I added a rudimentary fix.
image
This one I can reproduce locally, but unsure what's going on.

**Comment by 0101** [comment]
`TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject` this fails when running in parallel? 

That's suspicious, first thought would be either a shared instance of Checker or using the filesystem. Although neither of those should really be the case. Maybe the options cache but that looks like it should work too. Also if these were the problem other tests would probably be failing. Maybe you found an actual bug? 🤔

**Comment by majocha** [comment]
image
This one and all the rest in that file that are flaky are tricky to fix.
Generally, the tests here follow the schema of recording job events from the caches and comparing them to expected:
https://github.com/dotnet/fsharp/blob/d566f53680bf72ec0bec92d84ca2315896dd9266/tests/FSharp.Compiler.ComponentTests/FSharpChecker/TransparentCompiler.fs#L221-L243
It seems flakiness comes from a spurious `Finished` event that sometimes comes as the first one at the beginning.
That `Async.Sleep` 50 tries to filter it out but not always succeeds. We can just filter it out in the event stream but I'd like to understand where it comes from.

**Comment by majocha** [comment]
> `TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject` this fails when running in parallel?
> 
> That's suspicious, first thought would be either a shared instance of Checker or using the filesystem. Although neither of those should really be the case. Maybe the options cache but that looks like it should work too. Also if these were the problem other tests would probably be failing. Maybe you found an actual bug? 🤔

Now I'm not sure if it's not related to the changes I made here to graph processing, I'll try to run this particular test in the other branch to make sure.

**Comment by psfinaki** [comment]
Nice! Yeah I forgot about that, this aggressive mode can also help find some flaky stuff sometimes.

**Comment by majocha** [comment]
Also:
```
 FSharpChecker.TransparentCompiler.TryGetRecentCheckResultsForFile returns no result after edit
  No source available

Test has multiple result outcomes
   3 Passed
   1 Failed

Results

    Iteration 4)   FSharpChecker.TransparentCompiler.TryGetRecentCheckResultsForFile returns no result after edit 
      Duration: 1.1 sec

      Message: 
System.Exception : expected Some, but was None

      Stack Trace: 
ProjectOperations.expectSome[a](FSharpOption`1 x) line 779
Pipe #1 input at line 919@924-5.Invoke(FSharpOption`1 x)
TryGetRecentCheckResults@1381-3.Invoke(FSharpProjectSnapshot _arg43) line 1382
AsyncPrimitives.CallThenInvokeNoHijackCheck[a,b](AsyncActivation`1 ctxt, b result1, FSharpFunc`2 userCode) line 528
Trampoline.Execute(FSharpFunc`2 firstAction) line 112
--- End of stack trace from previous location ---
AsyncResult`1.Commit() line 454
AsyncPrimitives.QueueAsyncAndWaitForResultSynchronously[a](CancellationToken token, FSharpAsync`1 computation, FSharpOption`1 timeout) line 1143
AsyncPrimitives.RunSynchronously[T](CancellationToken cancellationToken, FSharpAsync`1 computation, FSharpOption`1 timeout) line 1170
FSharpAsync.RunSynchronously[T](FSharpAsync`1 computation, FSharpOption`1 timeout, FSharpOption`1 cancellationToken) line 1519
ProjectWorkflowBuilder.Execute(FSharpAsync`1 workflow) line 1039
ProjectWorkflowBuilder.Run(FSharpAsync`1 workflow) line 1050
TransparentCompiler.TryGetRecentCheckResultsForFile returns no result after edit() line 919
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

**Comment by majocha** [comment]
> > `TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject` this fails when running in parallel?
> > That's suspicious, first thought would be either a shared instance of Checker or using the filesystem. Although neither of those should really be the case. Maybe the options cache but that looks like it should work too. Also if these were the problem other tests would probably be failing. Maybe you found an actual bug? 🤔
> 
> Now I'm not sure if it's not related to the changes I made here to graph processing, I'll try to run this particular test in the other branch to make sure.

Nope, I can still repro:
```
 FSharpChecker.TransparentCompiler.TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject
  No source available

Test has multiple result outcomes
   8 Passed
   1 Failed

Results

    Iteration 9)   FSharpChecker.TransparentCompiler.TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject 
      Duration: 1.5 sec

      Message: 
System.Exception : expected Some, but was None

      Stack Trace: 
ProjectOperations.expectSome[a](FSharpOption`1 x) line 779
Pipe #1 input at line 910@911.Invoke(FSharpOption`1 x)
TryGetRecentCheckResults@1381-3.Invoke(FSharpProjectSnapshot _arg43) line 1382
AsyncPrimitives.CallThenInvokeNoHijackCheck[a,b](AsyncActivation`1 ctxt, b result1, FSharpFunc`2 userCode) line 528
Trampoline.Execute(FSharpFunc`2 firstAction) line 112
--- End of stack trace from previous location ---
AsyncResult`1.Commit() line 454
AsyncPrimitives.QueueAsyncAndWaitForResultSynchronously[a](CancellationToken token, FSharpAsync`1 computation, FSharpOption`1 timeout) line 1143
AsyncPrimitives.RunSynchronously[T](CancellationToken cancellationToken, FSharpAsync`1 computation, FSharpOption`1 timeout) line 1170
FSharpAsync.RunSynchronously[T](FSharpAsync`1 computation, FSharpOption`1 timeout, FSharpOption`1 cancellationToken) line 1519
ProjectWorkflowBuilder.Execute(FSharpAsync`1 workflow) line 1039
ProjectWorkflowBuilder.Run(FSharpAsync`1 workflow) line 1050
TransparentCompiler.TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject() line 910
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

I can repro this in just a couple iterations on an older intel laptop. A beefier 20 core i5 desktop I can run hundreds of iterations without hitting this.

**Comment by majocha** [comment]
Ok, I think I see where the bug is. `AsyncMemoize` does not do cache updates atomically but posts them with Task.Run. There is no guarantee of order of execution so under stress it's possible to go to `checkFile` in the workflow while the cache is not yet updated, despite the fact `Get` method returned. I'll try to verify this later.

**Comment by 0101** [comment]
> Ok, I think I see where the bug is. `AsyncMemoize` does not do cache updates atomically but posts them with `Task.Run`. There is no guarantee of order of execution so under stress it's possible to proceed to `checkFile` in the test workflow while the cache is not yet updated, despite the fact `Get` method returned. I'll try to verify this later.

`Get` should be atomic and "blocking". It should be fully completed when it returns (in terms of what's in the cache). Then when jobs either complete or are canceled that's sent via `Task.Run` whenever it happens and there is no defined order of execution. It's just serialized in some order by the lock. 

But maybe, under stress, by the time control was returned to whoever called `Get`, some further updates already happened on the requested job that were not expected? 🤔

**Comment by majocha** [comment]
> > Ok, I think I see where the bug is. `AsyncMemoize` does not do cache updates atomically but posts them with `Task.Run`. There is no guarantee of order of execution so under stress it's possible to proceed to `checkFile` in the test workflow while the cache is not yet updated, despite the fact `Get` method returned. I'll try to verify this later.
> 
> `Get` should be atomic and "blocking". It should be fully completed when it returns (in terms of what's in the cache). Then when jobs either complete or are canceled that's sent via `Task.Run` whenever it happens and there is no defined order of execution. It's just serialized in some order by the lock.
> 
> But maybe, under stress, by the time control was returned to whoever called `Get`, some further updates already happened on the requested job that were not expected? 🤔

Ah, yes you're right. I was looking at this line
https://github.com/dotnet/fsharp/blob/b34e992f02fe5de5d890cedd0621610edb45c3b0/src/Compiler/Facilities/AsyncMemoize.fs#L523
Which does a fire-and-forget task, but now I see the cache already has a value at this moment by earlier `GetOrCompute`.

**Comment by majocha** [comment]
> `TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject` this fails when running in parallel?
> 
> That's suspicious, first thought would be either a shared instance of Checker or using the filesystem. Although neither of those should really be the case. Maybe the options cache but that looks like it should work too. Also if these were the problem other tests would probably be failing. Maybe you found an actual bug? 🤔

This showed up again in CI. I'm thinking now it's a genuine bug, not just test case being flaky. Too bad I can only repro it locally on an underpowered laptop.

**Comment by majocha** [comment]
> > Ok, I think I see where the bug is. `AsyncMemoize` does not do cache updates atomically but posts them with `Task.Run`. There is no guarantee of order of execution so under stress it's possible to proceed to `checkFile` in the test workflow while the cache is not yet updated, despite the fact `Get` method returned. I'll try to verify this later.
> 
> `Get` should be atomic and "blocking". It should be fully completed when it returns (in terms of what's in the cache). Then when jobs either complete or are canceled that's sent via `Task.Run` whenever it happens and there is no defined order of execution. It's just serialized in some order by the lock.
> 
> But maybe, under stress, by the time control was returned to whoever called `Get`, some further updates already happened on the requested job that were not expected? 🤔

So, I was onto someithng after all. The test in question, `TryGetRecentCheckResultsForFile returns result after first call to ParseAndCheckFileInProject`
 hits `AsyncMemoize.TryGet` immediately after checking the project. `TryGet` scans the cache for `Completed` jobs.
The problem is that there are none yet, despite the fact previous `AsyncMemoize.Get` already finished. Why? because Get does not do `cache.Set`, it just queues it with a `Task.Run`. It allows a state where the overall async computation continues but the cache is not keeping up.

Ideally, the moment `Get` completes the cache should be up to date with the current state of the given job.

**Comment by psfinaki** [comment]
Thanks for the further investigations :)

Refactoring is okay as long as it is justified and not too crazy.

**Comment by majocha** [comment]
> Refactoring is okay as long as it is justified and not too crazy.

Here's the gist of it: https://gist.github.com/majocha/6c075dcdabf45c3fd92b03e2cf3a07eb

**Comment by 0101** [comment]
Good catch @majocha ! 

I don't know if this is a real issue in practice though. In this situation I think it's ok to say we don't have recent results yet. Alternatively it means some extra delay for the original caller waiting for cache update to be processed.

If it's used in the usual async way, you'd get back the `Running` job which will be completed very soon.

Unless I'm missing some other scenarios where this could be an issue, we could just update the test (maybe wait for cache event or something).

I didn't study your refactoring in depth yet but it's much less code, so if it does the same thing I'm all for it! But maybe could be in a separate PR.

**Comment by majocha** [comment]
> Good catch @majocha !
> 
> I don't know if this is a real issue in practice though. In this situation I think it's ok to say we don't have recent results yet. Alternatively it means some extra delay for the original caller waiting for cache update to be processed.
> 
> If it's used in the usual async way, you'd get back the `Running` job which will be completed very soon.
> 
> Unless I'm missing some other scenarios where this could be an issue, we could just update the test (maybe wait for cache event or something).

Yes, in practice fixing the particular test would be just to add a wait for `Completed` event.

> I didn't study your refactoring in depth yet but it's much less code, so if it does the same thing I'm all for it! But maybe could be in a separate PR.

That's what I'm thinking, first I want to get the tests up and running again.

**Comment by majocha** [comment]
> Unless I'm missing some other scenarios where this could be an issue, we could just update the test (maybe wait for cache event or something).

This actually cleared a huge blindspot for me. The whole thing I've been fighting here regarding the flakiness of these tests is that events can occur with some unpredictable delay. I did a lot of "clever" stuff here awaiting on the events but now it occurred to me:
A bool prop is needed on the `AyncMemoize` that indicates it has updates "in flight".
Then we can simply do a spinwait to ensure it is in a consistent state before asserting anything about the recorded events.

**Comment by 0101** [comment]
Good idea! Wonder if it takes care of all situations. It just might!

We could potentially also consider waiting until there are 0 `Running` jobs.

**Comment by majocha** [comment]
> Good idea! Wonder if it takes care of all situations. It just might!
Yes it does, I just redid the AsyncMemoize tests also with it. Now I should probably cross the test cases with their historical versions. They kinda went full circle and back with all the changes as I was fighting (having fun) with it.

**Comment by majocha** [comment]
> We could potentially also consider waiting until there are 0 `Running` jobs.

Yes, I considered this, but the cache state only gives no guarantee that there is no pending state update queued in the thread pool.

**Comment by 0101** [comment]
> > We could potentially also consider waiting until there are 0 `Running` jobs.
> 
> Yes, I considered this, but the cache state only gives no guarantee that there is no pending state update queued in the thread pool.

Well yes, but also no state update (coming through `post`) can increase the number of `Running` jobs.

**Comment by majocha** [comment]
This is now cherrypicked as PR to main #17966

**Comment by T-Gro** [comment]
Thanks Jakub ==> I am closing this one then.

**Pull request closed without merging**